### PR TITLE
store: Be pessimistic and check how many bytes got copied

### PR DESCRIPTION
**Pull request by zecke**
I have seen a panic in loadChunks and the first thing that came to mind
that the getRange query might not give the result we would like to have.
Check the right number of bytes were copied into the destination.

Signed-off-by: Holger Hans Peter Freyther <holger@freyther.de>

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Verification

Not yet. Soliciting feedback of how defensive we want to be.

### CODE REVIEW

**Comment by pracucci** [comment]
Do you have a trace of the panic you're experiencing? I would be glad to learn more about the issue you're experiencing.

**Comment by zecke** [comment]
> Do you have a trace of the panic you're experiencing? I would be glad to learn more about the issue you're experiencing.

The panic trace looks like the one here: https://github.com/thanos-io/thanos/issues/3339

I added a defer + recover in there so that a single failure doesn't crash the store and will add enough debug information to find the block. My two cents are I would like this part of the code be defensive:

* Don't crash and take the process down (is it a corrupted block? is it something funny the object store in use)
* Be defensive. E.g. io.CopyN and check that we have copied that many bytes..

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -1472,10 +1472,15 @@ func (b *bucketBlock) readChunkRange(ctx context.Context, seq int, off, length i
...
 
-	if _, err = io.Copy(buf, r); err != nil {
+	copied, err := io.Copy(buf, r)
+	if err != nil {
 		b.chunkPool.Put(c)
 		return nil, errors.Wrap(err, "read range")
 	}
+	if copied != len {
+		b.chunkPool.Put(c)
+		return nil, errors.Errorf("Read less than specified wanted=%v got=%v", len, copied)
```
Comment (GiedriusS) [inline pkg/store/bucket.go:1482]: We should begin it with a lowercase letter. Also, I think this improves the formatting:

```suggestion
		return nil, errors.Errorf("read %v bytes but got %v", len, copied)
```

**Review by GiedriusS — COMMENTED** [review]
I think this change is reasonable. One minor nit on the error message. Care to update this PR to fix the build errors?

File: pkg/store/bucket.go [hunk 2]
```diff
@@ -1472,10 +1472,15 @@ func (b *bucketBlock) readChunkRange(ctx context.Context, seq int, off, length i
 	}
 	defer runutil.CloseWithLogOnErr(b.logger, r, "readChunkRange close range reader")
 
-	if _, err = io.Copy(buf, r); err != nil {
+	copied, err := io.Copy(buf, r)
+	if err != nil {
 		b.chunkPool.Put(c)
 		return nil, errors.Wrap(err, "read range")
 	}
+	if copied != len {
```
Comment (pracucci) [inline pkg/store/bucket.go:1480]: I guess you meant `if copied != length`

Comment (zecke) [reply pkg/store/bucket.go:1480]: doh. Thanks linter. ;)

**Comment by zecke** [comment]
The assumption I made about the code and reality do not align yet. My assumption was that we precisely  know where the data we want is located and how large it is. Let me run this locally to understand where the gap is. Any ideas/pointers welcome.

```
level=error ts=2020-12-22T11:20:17.1040842Z msg="function failed. Retrying in next tick" err="unexpected warnings [receive series from Addr: e2e_test_compact-store-gw-1:9091 LabelSets: {case=\"a-partial-overlap-dedup-ready\", replica=\"1\"},{case=\"a-partial-overlap-dedup-ready\", replica=\"2\"},{case=\"a-partial-overlap-dedup-ready\", replica=\"3\"},{case=\"compaction-ready\", replica=\"1\"},{case=\"compaction-ready-after-dedup\", replica=\"1\"},{case=\"compaction-ready-after-dedup\", replica=\"2\"},{case=\"compaction-ready-one-block-marked-for-no-compact\", replica=\"1\"},{case=\"full-replica-overlap-dedup-ready\", replica=\"1\"},{case=\"full-replica-overlap-dedup-ready\", replica=\"2\"},{case=\"no-compaction\", replica=\"1\"},{case=\"partial-multi-replica-overlap-dedup-ready\", replica=\"1\", rule_replica=\"1\"},{case=\"partial-multi-replica-overlap-dedup-ready\", replica=\"1\", rule_replica=\"2\"} Mint: 1585036800000 Maxt: 1585072800000: rpc error: code = Aborted desc = fetch series for block 01ET510R8RSYN56Z5THSV4V2SP: preload chunks: read range for 0: read 17798 bytes but got 1839]"
```

**Pull request closed without merging**